client/src/GameState.ts

```diff
@@ -126,6 +205,157 @@ export class GameState {
+                            newMonomino.position[0] = newRow;
+                        }
+                    }
+                } else if (task.axis === LineCheckAxis.Column) {
+                    const offset =
+                        task.direction * (linesToFall.get(lineIndex) || 0);
+                    if (offset === 0) {
+                        return;
+                    }
+                    const newCol = lineIndex + offset;
+
+                    for (let row = 15; row < 25; row++) {
+                        this.board[row][newCol] = this.board[row][lineIndex];
+                        this.board[row][lineIndex] = null;
+                        const newMonomino = this.board[row][newCol];
+                        if (newMonomino) {
+                            newMonomino.position[1] = newCol;
+                        }
+                    }
+                }
+            });
+    }
+
+    private prepareLinesToFall(
+        task: LineCheckTask,
+        linesToClear: Array<number>
+    ): Map<number, number> {
+        const linesOffset = new Map<number, number>();
+        let offset = 0;
+        if (linesToClear.length > 0) {
+            task.range
+                .slice()
+                .reverse()
+                .forEach((lineIndex) => {
+                    if (lineIndex === linesToClear?.at(0)) {
+                        // this line is an empty, cleared line. all lines above should fall with one more offset
+                        offset += 1;
+                        linesToClear.push(linesToClear.shift()!);
+                    } else {
+                        linesOffset.set(lineIndex, offset);
+                    }
+                });
+        }
+        return linesOffset;
+    }
+
+    private scanLinesToClear(task: LineCheckTask): Array<number> {
+        const linesToClear = [];
+        if (task.axis === LineCheckAxis.Row) {
+            for (const row of task.range) {
+                // scan through each row
+                let canClear = true;
+                for (let col = 15; col < 25; col++) {
+                    if (this.board && this.board[row][col] === null) {
+                        canClear = false;
+                        break;
+                    }
+                }
+                if (canClear) {
+                    linesToClear.push(row);
+                }
+            }
+        } else if (task.axis === LineCheckAxis.Column) {
+            for (const col of task.range) {
+                // scan through each col
+                let canClear = true;
+                for (let row = 15; row < 25; row++) {
+                    if (this.board && this.board[row][col] === null) {
+                        canClear = false;
+                        break;
+                    }
+                }
+                if (canClear) {
+                    linesToClear.push(col);
+                }
+            }
+        }
+
+        return linesToClear.reverse();
```

Comment:
Is the reverse needed?

Reply:
I think yes because we need to scan through the lines to clear in the bottom-up order, so that the other lines can know how many blocks to fall after clearing.